File: minWindowSubstr.cpp
/* Minimum window substring (99):
Given a string S and a string T, find the minimum window in S which will 
contain all the characters in T in complexity O(n).
For example, S = "ADOBECODEBANC", T = "ABC", Minimum window is "BANC".
http://www.programcreek.com/2014/05/leetcode-minimum-window-substring-java/ */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

using namespace std;

string getMinWindow(const string &s, const string &t)
{
	unordered_map<char, int> map;
	for (int i = 0; i < t.length(); ++i) {
		map[t[i]] = 0;
	}
	
	int start = 0, bestStart = -1, minLen = INT_MAX;
	unordered_set<int> set;
	for (int i = 0; i < s.length(); ++i) {
		if (map.count(s[i]) > 0 && set.count(s[i]) == 0) {
			set.insert(s[i]);
			++map[s[i]];
			
			while (set.size() == map.size()) {
				if (minLen > (i - start + 1)) {
					minLen = i - start + 1;
					bestStart = start;
				}
				
				--map[s[start]];
				if (map[s[start]] == 0) {
					set.erase(s[start]);
				}
				
				++start;
				while (start < i && map.count(s[start]) == 0) {
					++start;
				}
				
				if (start >= i) {
					break;
				}
			}
		} else if (set.count(s[i]) > 0) {
			++map[s[i]];
		}
	}
	
	return s.substr(bestStart, minLen);
}

int main(int argc, char **argv)
{
	if (argc < 3) {
		cout << "Usage: a.out <S> <T>" << endl;
		return 1;
	}
	
	string s(argv[1]);
	string t(argv[2]);
	cout << "S: " << s << ", t: " << t << endl;
	string minWindow = getMinWindow(s, t);
	cout << "Min window: " << minWindow << endl;
	 
	return 0;
}

File: stringsArrays/MinWindowSubstr.java
/**
 * Minimum window substring (99):
 * Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
 * For example, S = "ADOBECODEBANC", T = "ABC", Minimum window is "BANC".
 * http://www.programcreek.com/2014/05/leetcode-minimum-window-substring-java/
 */

import java.util.*;

public class MinWindowSubstr {
    public String minWindow(String s, String t) {
        int sl = s.length();
        int tl = t.length();
        String result = "";
        if (sl < tl) {
            return result;
        }

        // refmap
        Map<Character, Integer> refMap = new HashMap<>();
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < tl; ++i) {
            char c = t.charAt(i);
            refMap.put(c, refMap.getOrDefault(c, 0) + 1);
            map.put(c, 0);
        }

        int i = 0, j = 0, min = Integer.MAX_VALUE, count = 0;
        Deque<Integer> q = new ArrayDeque<>();
        while (j < sl) {
            char c = s.charAt(j);
            if (refMap.containsKey(c)) {
                if (map.get(c) < refMap.get(c)) {
                    ++count;
                }
                map.put(c, map.getOrDefault(c, 0) + 1);
                q.addLast(j);
            }
            ++j;

            if (count < tl) {
                continue;
            }

            if (min > j - i) {
                min = j - i;
                result = s.substring(i, j);
            }

            while (q.size() > 0 && count == tl) {
                i = q.removeFirst();
                char d = s.charAt(i);
                if (map.get(d) <= refMap.get(d)) {
                    --count;
                }

                if (min > j - i) {
                    min = j - i;
                    result = s.substring(i, j);
                }
                map.put(d, map.get(d) - 1);
            }
        }

        return result;
    }

    public static void main(String[] args) {
        MinWindowSubstr mws = new MinWindowSubstr();
        System.out.println("s: " + args[0] + ", t: " + args[1]);
        String result = mws.minWindow(args[0], args[1]);
        System.out.println("result: " + result);
    }
}
